doc: nrf: drivers: Add documentation for sQSPI MSPI shim driver (#22029)
nrfconnect/sdk-nrf: doc/nrf/drivers/mspi_sqspi.rst

					<NRF_PSEL(SDP_MSPI_DQ0, 7, 1)>,
					<NRF_PSEL(SDP_MSPI_DQ1, 7, 2)>,
					<NRF_PSEL(SDP_MSPI_DQ2, 7, 3)>,
					<NRF_PSEL(SDP_MSPI_DQ3, 7, 4)>;
			};
		};
	};

	&cpuflpr_vpr {
		pinctrl-0 = <&sqspi_default>;
		pinctrl-1 = <&sqspi_sleep>;
		pinctrl-names = "default", "sleep";
		interrupts = <212 NRF_DEFAULT_IRQ_PRIORITY>;
		status = "okay";
		/delete-property/ execution-memory;
	};

	/delete-node/ &cpuflpr_code_data;
	/delete-node/ &cpuapp_cpuflpr_ipc_shm;
	/delete-node/ &cpuflpr_cpuapp_ipc_shm;
	/delete-node/ &cpuapp_cpuflpr_ipc;

	&ram21_region {
		status = "okay";

		softperiph_ram: memory@0 {
			reg = <0 0x4000>;
			ranges;
			#address-cells = <1>;
			#size-cells = <1>;

			sqspi: sqspi@3e00 {
				compatible = "nordic,nrf-sqspi";
				#address-cells = <1>;
				#size-cells = <0>;
				reg = <0x3e00 0x200>;
				zephyr,pm-device-runtime-auto;
				memory-regions = <&sqspi_buffers>;
			};
		};

		sqspi_buffers: memory@4000 {
			compatible = "zephyr,memory-region";
			reg = <0x4000 0x4000>;
			#memory-region-cells = <0>;
			zephyr,memory-region = "SQSPI_BUFFERS";
			zephyr,memory-attr = <DT_MEM_CACHEABLE>;
		};
	};

API documentation
*****************

> 
> | Source file: :file:`drivers/mspi/mspi_sqspi.c`

Reviewer: ```suggestion

| Header file: :file:`include/zephyr/drivers/mspi.h`
| Source file: :file:`drivers/mspi/mspi_sqspi.c`
```

.. _mspi_sqspi:

sQSPI MSPI shim driver
######################

.. contents::
   :local:
   :depth: 2

This driver integrates the :ref:`sQSPI` soft peripheral with the Zephyr MSPI API, enabling control over the peripheral.
This allows for communication with devices that use MSPI bus-based Zephyr drivers, such as flash chips or displays.

Configuration
*************

> You must configure the sQSPI MSPI shim driver with the devicetree:

Reviewer 2: ```suggestion
To configure the sQSPI MSPI shim driver with the devicetree, complete the following steps :
```

1. Specify the base address of the sQSPI registers in RAM.

>    You can do this by creating the :dtcompatible:`nordic,nrf-sqspi` node within the RAM region reserved for the soft peripheral.

Reviewer 2: ```suggestion
   Create the :dtcompatible:`nordic,nrf-sqspi` node within the RAM region reserved for the soft peripheral.
```

Author: How about:
```
1. Specify the base address of the sQSPI registers in RAM by creating the :dtcompatible:`nordic,nrf-sqspi` node within the RAM region reserved for the soft peripheral.
```
This is one thing. The user by creating the node and setting its base address indicates to the shim driver where to access the sQSPI registers.

.. _mspi_sqspi:

sQSPI MSPI shim driver
######################

.. contents::
   :local:
   :depth: 2

> This driver provides an integration layer for the :ref:`sQSPI` soft peripheral
> that makes it possible to control the peripheral with the Zephyr MSPI API.
> This in turn allows using it for communication with any devices for which there
> are MSPI bus based Zephyr drivers available, like flash chips or displays.

Reviewer: ```suggestion
This driver integrates the :ref:`sQSPI` soft peripheral with the Zephyr MSPI API, enabling control over the peripheral.
This allows for communication with devices that use MSPI bus-based Zephyr drivers, such as flash chips or displays.
```

Configuration
*************

> The sQSPI MSPI shim driver is configured through the devicetree. It must be
> provided with the base address of the sQSPI registers in RAM. This is done
> by defining a :dtcompatible:`nordic,nrf-sqspi` node within the RAM region
> reserved for the soft peripheral. Additionally, an interrupt for communication
> between the Application and FLPR cores needs to be specified, as well as pins
> on which the sQSPI is to operate. These two things are supplied in properties
> of the node labelled ``cpuflpr_vpr``.
> It is also important to delete the ``execution-memory`` property in the
> ``cpuflpr_vpr`` node if it exists, so that the Nordic VPR coprocessor launcher
> is not used for the FLPR core (see :kconfig:option:`CONFIG_NORDIC_VPR_LAUNCHER`).
> 
> Details on RAM configuration for the SoCs supported by sQSPI can be found in:
> - ::ref:`nrf54L15_porting_guide_ram_configuration`
> - ::ref:`nrf54H20_porting_guide_ram_configuration`
> 
> The RAM region reserved for the sQSPI needs to be labelled ``softperiph_ram``
> on nRF54H20, so that it can be set as non-cacheable in the MPU configuration.
> 
> See the following configuration example for the nRF54L15 SoC:

Reviewer: ```suggestion
To configure the sQSPI MSPI shim driver, use the devicetree and complete the following steps:

1. Define a node by creating the :dtcompatible:`nordic,nrf-sqspi` node within the RAM region reserved for the soft peripheral.
#. Specify an interrupt for communication between the application and FLPR cores, and determine the pins for sQSPI operation.
  You can find this information in the properties of the node labeled as ``cpuflpr_vpr``.
#. If the ``execution-memory`` property exists in the ``cpuflpr_vpr`` node, delete it to ensure the Nordic VPR coprocessor launcher does not use the FLPR core (see the :kconfig:option:`CONFIG_NORDIC_VPR_LAUNCHER` Kconfig option).

For RAM configuration details on supported SoCs, refer to the following guides:

* :ref:`nRF54L15 RAM configuration <nrf54L15_porting_guide_ram_configuration>`
* :ref:`nRF54H20 RAM configuration <nrf54H20_porting_guide_ram_configuration>` - On the nRF54H20 SoC, ensure the RAM region reserved for the sQSPI is labeled ``softperiph_ram`` to set it as non-cacheable in the MPU configuration.
  
See the following configuration example for the nRF54L15 SoC:
```

			};
			group2 {
				psels = <NRF_PSEL(SDP_MSPI_DQ1, 2, 4)>,
					<NRF_PSEL(SDP_MSPI_DQ2, 2, 3)>,
					<NRF_PSEL(SDP_MSPI_DQ3, 2, 0)>;
				nordic,drive-mode = <NRF_DRIVE_E0E1>;
				bias-pull-up;
			};
		};

		sqspi_sleep: sqspi_sleep {
			group1 {
				low-power-enable;
				psels = <NRF_PSEL(SDP_MSPI_SCK, 2, 1)>,
					<NRF_PSEL(SDP_MSPI_CS0, 2, 5)>,
					<NRF_PSEL(SDP_MSPI_DQ0, 2, 2)>,
					<NRF_PSEL(SDP_MSPI_DQ1, 2, 4)>,
					<NRF_PSEL(SDP_MSPI_DQ2, 2, 3)>,
					<NRF_PSEL(SDP_MSPI_DQ3, 2, 0)>;
			};
		};
	};

	&cpuflpr_vpr {
		pinctrl-0 = <&sqspi_default>;
		pinctrl-1 = <&sqspi_sleep>;
		pinctrl-names = "default", "sleep";
		interrupts = <76 NRF_DEFAULT_IRQ_PRIORITY>;
		status = "okay";
	};

	/ {
		reserved-memory {
			#address-cells = <1>;
			#size-cells = <1>;
			ranges;

			softperiph_ram: memory@2003c000 {
				reg = <0x2003c000 0x4000>;
				ranges = <0 0x2003c000 0x4000>;
				#address-cells = <1>;
				#size-cells = <1>;

				sqspi: sqspi@3c00 {
					compatible = "nordic,nrf-sqspi";
					#address-cells = <1>;
					#size-cells = <0>;
					reg = <0x3c00 0x200>;
					status = "okay";
					zephyr,pm-device-runtime-auto;
				};
			};
		};
	};

> For the nRF54H20 SoC, additionally a region for data buffers to be used in
> sQSPI transfers must be reserved in RAM accessible by the FLPR core. This
> region needs to be indicated to the shim driver by the ``memory-regions``
> property in the node representing the sQSPI. If an MSPI transfer will be
> requested with a buffer located outside of this region, the shim driver will
> allocate a buffer in the region for the time of the transfer and do the proper
> copying of data between these two buffers (before a TX transfer or after an RX
> transfer). To avoid the overhead of this automatic allocation and copying,
> a proper buffer can be allocated statically using :c:macro:`DMM_MEMORY_SECTION`.
> Then the shim driver will pass such buffer directly to the sQSPI.

Reviewer: ```suggestion
For the nRF54H20 SoC, you must also reserve a RAM region for data buffers used in
sQSPI transfers, accessible by the FLPR core. 
Indicate this region to the shim driver using the ``memory-regions`` property in the sQSPI node. 
If you initiate an MSPI transfer with a buffer outside this region, the shim driver will temporarily allocate a buffer within the region and correctly transfer data between the two buffers - before transmitting (TX) or after receiving (RX).
To avoid overhead from automatic allocation and copying, allocate buffers statically using :c:macro:`DMM_MEMORY_SECTION`. 
The shim driver will then directly pass these buffers to the sQSPI.
```

> An example configuration for the nRF54H20 SoC:

Reviewer: ```suggestion
The following example configuration for the nRF54H20 SoC sets up the necessary parameters:
```

					<NRF_PSEL(SDP_MSPI_CS0, 7, 5)>,
					<NRF_PSEL(SDP_MSPI_DQ0, 7, 1)>,
					<NRF_PSEL(SDP_MSPI_DQ1, 7, 2)>,
					<NRF_PSEL(SDP_MSPI_DQ2, 7, 3)>,
					<NRF_PSEL(SDP_MSPI_DQ3, 7, 4)>;
			};
		};
	};

	&cpuflpr_vpr {
		pinctrl-0 = <&sqspi_default>;
		pinctrl-1 = <&sqspi_sleep>;
		pinctrl-names = "default", "sleep";
		interrupts = <212 NRF_DEFAULT_IRQ_PRIORITY>;
		status = "okay";
		/delete-property/ execution-memory;
	};

	/delete-node/ &cpuflpr_code_data;
	/delete-node/ &cpuapp_cpuflpr_ipc_shm;
	/delete-node/ &cpuflpr_cpuapp_ipc_shm;
	/delete-node/ &cpuapp_cpuflpr_ipc;

	&ram21_region {
		status = "okay";

		softperiph_ram: memory@0 {
			reg = <0 0x4000>;
			ranges;
			#address-cells = <1>;
			#size-cells = <1>;

			sqspi: sqspi@3e00 {
				compatible = "nordic,nrf-sqspi";
				#address-cells = <1>;
				#size-cells = <0>;
				reg = <0x3e00 0x200>;
				zephyr,pm-device-runtime-auto;
				memory-regions = <&sqspi_buffers>;
			};
		};

		sqspi_buffers: memory@4000 {
			compatible = "zephyr,memory-region";
			reg = <0x4000 0x4000>;
			#memory-region-cells = <0>;
			zephyr,memory-region = "SQSPI_BUFFERS";
			zephyr,memory-attr = <DT_MEM_CACHEABLE>;
		};
	};

API Reference
*************

> 

Reviewer: Is there any header/source file path that can be added here? (example: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/drivers/sensor_stub.html#api_documentation )

Author: Added a source file.
